File: scenarios/command/dotnet/command_consumer/CommandClient.cs
using MQTTnet;
using MQTTnet.Client;
using proto_messages;

namespace command_consumer;

internal class CommandClient<T, TResp>
{
    readonly IMqttClient _mqttClient;
    readonly string _commandName;
    string _remoteClientId = string.Empty;
    TaskCompletionSource<TResp>? _tcs;
    IMessageSerializer _serializer;
    Guid _correlationId;

    public string RequestTopicPattern { get; set; } = "device/{clientId}/command/{commandName}/request";
    public string ResponseTopicPattern { get; set; } = "device/{clientId}/command/{commandName}/response";

    public CommandClient(IMqttClient mqttClient, string cmdName)
    {
        _mqttClient = mqttClient;
        _commandName = cmdName;
        _serializer = new ProtobufSerializer(unlockResponse.Parser);
        _mqttClient.ApplicationMessageReceivedAsync += async m =>
        {
            var topic = m.ApplicationMessage.Topic;
            var expectedTopic = ResponseTopicPattern.Replace("{clientId}", _remoteClientId).Replace("{commandName}", cmdName);
            if (topic.Equals(expectedTopic))
            {
                if (m.ApplicationMessage.CorrelationData != null && _correlationId != new Guid(m.ApplicationMessage.CorrelationData))
                {
                    _tcs!.SetException(new ApplicationException("Invalid correlation data"));
                }
                else
                {
                    TResp response = _serializer.FromBytes<TResp>(m.ApplicationMessage.Payload);
                    _tcs!.SetResult(response);
                }
            }
            await Task.Yield();
        };
    }
    public async Task<TResp> InvokeAsync(string clientId, T request, CancellationToken ct = default)
    {
        _remoteClientId = clientId;
        _correlationId = Guid.NewGuid();
        string requestTopic = RequestTopicPattern!.Replace("{clientId}", clientId).Replace("{commandName}", _commandName);
        string responseTopic = ResponseTopicPattern.Replace("{clientId}", clientId).Replace("{commandName}", _commandName);
        _tcs = new TaskCompletionSource<TResp>();
        await _mqttClient.SubscribeAsync(responseTopic);
        await _mqttClient.PublishAsync(new MqttApplicationMessageBuilder()
            .WithTopic(requestTopic)
            .WithPayload(_serializer.ToBytes(request!))
            .WithCorrelationData(_correlationId.ToByteArray())
            .Build());
        return await _tcs.Task;
    }
}

File: mqttclients/ts/mqttjsClientExtensions/src/commandClient.ts
import {
    IPublishPacket,
    MqttClient
} from 'mqtt';
import { v4 as uuidV4 } from 'uuid';
import {
    logger,
    DeferredPromise,
    IMessageSerializer
} from './';

const ModuleName = 'CommandClient';

export class CommandClient<T, TResp>
{
    private mqttClient: MqttClient;
    private commandName: string;
    private requestTopicPattern: string;
    private requestTopic: string;
    private responseTopicPattern: string;
    private responseTopic: string;
    private serializer: IMessageSerializer;
    private correlationId: string;
    private deferredPromise: DeferredPromise<TResp>;

    constructor(mqttClient: MqttClient, requestTopicPattern: string, responseTopicPattern: string, commandName: string, serializer: IMessageSerializer) {
        this.mqttClient = mqttClient;
        this.commandName = commandName;
        this.requestTopicPattern = requestTopicPattern;
        this.responseTopicPattern = responseTopicPattern;
        this.serializer = serializer;

        this.mqttClient.on('message', (topic: string, payload: Buffer, packet: IPublishPacket) => {
            if (topic === this.responseTopic) {
                if (packet.properties?.contentType !== serializer.contentType) {
                    logger.error({ tags: [ModuleName] }, `Message received on topic ${topic} but with invalid content type. Expected ${this.serializer.contentType} - received ${packet.properties?.contentType}`);
                }

                const responseCorrelationId = (packet.properties?.correlationData ?? '').toString();

                if (responseCorrelationId !== this.correlationId) {
                    logger.error({ tags: [ModuleName] }, `Message received on topic ${topic} but correlationId does not match. Expected ${this.correlationId} - received ${responseCorrelationId}`);
                }

                logger.info({ tags: [ModuleName] }, `Message received on topic: ${topic} with correlationId: ${responseCorrelationId}`);

                const response = this.serializer.fromBytes<TResp>(payload);

                return this.deferredPromise.resolve(response);
            }
        });
    }

    public async invokeAsync(clientId: string, request: T, timeoutInMilliSeconds = 5000): Promise<TResp> {
        this.requestTopic = this.requestTopicPattern.replace('{clientId}', clientId).replace('{commandName}', this.commandName);
        this.responseTopic = this.responseTopicPattern.replace('{clientId}', clientId).replace('{commandName}', this.commandName);

        this.correlationId = uuidV4();

        await this.mqttClient.subscribeAsync(this.responseTopic, {
            qos: 1,
        });

        const respBytes = this.serializer.toBytes(request);
        const pubAck = await this.mqttClient.publishAsync(this.requestTopic, respBytes, {
            qos: 1,
            properties: {
                contentType: this.serializer.contentType,
                correlationData: Buffer.from(this.correlationId),
                responseTopic: this.responseTopic,
                userProperties: {
                    status: '200'
                }
            }
        });

        logger.info({ tags: [ModuleName] }, `Published command request on topic: ${this.requestTopic}, with mid: ${pubAck?.messageId}, correlationId: ${this.correlationId}`);

        this.deferredPromise = new DeferredPromise<TResp>();
        return this.promiseWithTimeout<TResp>(this.deferredPromise.promise, timeoutInMilliSeconds);
    }

    private async promiseWithTimeout<T>(promise: Promise<T>, timeoutInMilliSeconds: number): Promise<T> {
        const timeout = new Promise<never>((_, reject) => {
            setTimeout(() => {
                reject(new Error(`Command response timed out after ${timeoutInMilliSeconds} milliseconds`));
            }, timeoutInMilliSeconds);
        });

        return Promise.race<T>([promise, timeout]);
    }
}
